rust/src/csvreader.rs

```diff
@@ -0,0 +1,243 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! CSV Support
+
+use std::fs::File;
+use std::io::BufReader;
+use std::sync::Arc;
+
+use array::ArrayRef;
+use builder::{ArrayBuilder, ListArrayBuilder, PrimitiveArrayBuilder};
+use datatypes::{DataType, Schema};
+use error::ArrowError;
+use record_batch::RecordBatch;
+
+use csv;
+use csv::{StringRecord, StringRecordsIntoIter};
+
+pub struct CsvFile {
+    schema: Arc<Schema>,
+    projection: Option<Vec<usize>>,
+    record_iter: StringRecordsIntoIter<BufReader<File>>,
+    batch_size: usize,
+}
+
+impl CsvFile {
+    pub fn open(
+        file: File,
+        schema: Arc<Schema>,
+        has_headers: bool,
+        batch_size: usize,
+        projection: Option<Vec<usize>>,
+    ) -> Self {
+        let csv_reader = csv::ReaderBuilder::new()
+            .has_headers(has_headers)
+            .from_reader(BufReader::new(file));
+
+        let record_iter = csv_reader.into_records();
+        CsvFile {
+            schema: schema.clone(),
+            projection,
+            record_iter,
+            batch_size,
+        }
+    }
+}
+
+macro_rules! build_primitive_array {
+    ($ROWS:expr, $COL_INDEX:expr, $TY:ty) => {{
+        let mut builder = PrimitiveArrayBuilder::<$TY>::new($ROWS.len() as i64);
+        for row_index in 0..$ROWS.len() {
+            match $ROWS[row_index].get(*$COL_INDEX) {
+                Some(s) if s.len() > 0 => builder.push(s.parse::<$TY>().unwrap()).unwrap(),
+                _ => builder.push_null().unwrap(),
+            }
+        }
+        Ok(Arc::new(builder.finish()) as ArrayRef)
+    }};
+}
+
+impl CsvFile {
+    /// Read the next batch of rows
+    pub fn next(&mut self) -> Option<Result<Arc<RecordBatch>, ArrowError>> {
+        // read a batch of rows into memory
+        let mut rows: Vec<StringRecord> = Vec::with_capacity(self.batch_size);
+        for _ in 0..self.batch_size {
+            match self.record_iter.next() {
```

Comment:
Should we expose `record_iter` to the user?  This would allow them to skip records at the beginning of a file for example before reading the rest of the file.

Reply:
I will think about this. There is already the constructor param to indicate whether there is a header row or not that needs to be skipped. Maybe we could do this as a separate enhancement PR.

Reply:
I have a use case where there is meta-data inserted after the header but before the data and I have no way to control this.  Perhaps this is a rather narrow use case.  In any case, happy to revisit this.